### PR TITLE
syncer(dm): prepare promutil.Factory injection and cache metrics

### PR DESCRIPTION
**Pull request by lance6716**
### What problem does this PR solve?

Issue Number: ref #5493

### What is changed and how it works?

syncer will check the injected Factory to create new metrics. And cache some metrics to avoid calling some heavy functions in hot path.

#### Tests 

 - pass original test. Test about injected Factory will come in future PR.
 - benchbot (86s vs 115s) origin_img_v2_f6898c67-0411-442c-9c1b-2bd350f3494g

### Release note 

```release-note
 `None`.
```

### CODE REVIEW

**Comment by lance6716** [comment]
PTAL @gozssky , similar to lightning's change

File: dm/syncer/dbconn/db.go [hunk 1]
```diff
@@ -84,7 +85,12 @@ func (conn *DBConn) ResetConn(tctx *tcontext.Context) error {
 }
 
 // QuerySQL does one query.
-func (conn *DBConn) QuerySQL(tctx *tcontext.Context, query string, args ...interface{}) (*sql.Rows, error) {
+func (conn *DBConn) QuerySQL(
+	tctx *tcontext.Context,
+	metricProxies *metrics.Proxies,
```
Comment (gozssky) [inline dm/syncer/dbconn/db.go:90]: Can we put `metricsProxies` in `tctx`?

Comment (lance6716) [reply dm/syncer/dbconn/db.go:90]: tcontext is a common package, not only for syncer's metrics. And some functions don't have a `tctx` argument, so I put metrics outside `tctx`

**Review by gozssky — APPROVED** [review]

File: dm/dm/config/subtask.go [hunk 2]
```diff
@@ -27,6 +27,7 @@ import (
 	"github.com/pingcap/tidb-tools/pkg/column-mapping"
 	extstorage "github.com/pingcap/tidb/br/pkg/storage"
 	"github.com/pingcap/tidb/util/filter"
+	"github.com/pingcap/tiflow/engine/pkg/promutil"
```
Comment (Ehco1996) [inline dm/dm/config/subtask.go:30]: this import order is wrong 🤔

File: dm/syncer/metrics/metrics.go [hunk 3]
```diff
@@ -30,302 +31,350 @@ const (
 	BinlogEventCostStageGenQuery      = "gen-query"
 )
 
-// below variables are exported to syncer package.
-var (
-	BinlogReadDurationHistogram = metricsproxy.NewHistogramVec(
+// Metrics groups syncer's metric variables.
+type Metrics struct {
```
Comment (Ehco1996) [inline dm/syncer/metrics/metrics.go:35]: how about renaming to `TaskMetrics`

File: dm/syncer/metrics/metrics.go [hunk 4]
```diff
@@ -30,302 +31,350 @@ const (
...
+	RemainingTimeGauge              prometheus.Gauge
+	ShardLockResolving              prometheus.Gauge
+	FinishedTransactionTotal        prometheus.Counter
+	FlushCheckPointsTimeInterval    prometheus.Observer
+}
+
+// Proxies provides the ability to clean Metrics values when syncer is closed.
+type Proxies struct {
+	Metrics                         *Metrics
+	BinlogReadDurationHistogram     *metricsproxy.HistogramVecProxy
```
Comment (Ehco1996) [inline dm/syncer/metrics/metrics.go:59]: it seems that only `Metrics` is accessed by callers, maybe make other members like  `BinlogReadDurationHistogram` to private members?

**Review by Ehco1996 — COMMENTED** [review]
rest LGTM

File: dm/syncer/metrics/metrics.go (hunk 3 above)
Comment (lance6716) [reply dm/syncer/metrics/metrics.go:35]: some metrics are not a property of a single task, for example BinlogMasterPosGauge. Welcome to give a better name 😂

File: dm/syncer/metrics/metrics.go (hunk 4 above)
Comment (lance6716) [reply dm/syncer/metrics/metrics.go:59]: some variables like UnsyncedTableGauge are accessed in syncer package, I'll make rest of them private.

File: dm/syncer/metrics/metrics.go (hunk 3 above)
Comment (Ehco1996) [reply dm/syncer/metrics/metrics.go:35]: In this case I do not have a good name too, the name is too difficult 😭

**Review by Ehco1996 — APPROVED** [review]
LGTM

**Pull request merged**